CBD-6474 JwtAuthenticator

dnault opened the pull request:
Describe a new authenticator that supports JSON Web Token (JWT) credentials.

Also: Refine the pseudocode for PasswordAuthenticator to reflect the actual value of the HTTP `Authorization` header.

dnault committed:
CBD-6474 JwtAuthenticator

Describe a new authenticator that supports
JSON Web Token (JWT) credentials.

dnault committed:
SDK must respond to status code 0x1f (auth stale)
by closing the connection.

dnault commented:
Update:
* SDK must respond to "auth stale" status code by closing the connection.

chvck commented on rfc/0048-sdk3-bootstrapping.md:
```diff
@@ -153,13 +153,23 @@ interface Authenticator {
+
+#### Stale credentials
+
+If a KV response has status code `0x1f` ("auth stale"), the SDK must close the connection that received the response, and replace it with a new connection.
```
What's the actual behaviour here? Can this be returned as a response to any request? Is the intention that the connection will get closed and the request is always retried? If so with any specific retry reason?

dnault replied:
> **Can this be returned as a response to any request?**

Yes.

> **Is the intention that the connection will get closed and the request is always retried?**

~Yes.~ The connection gets closed, and the request fails.

> **If so with any specific retry reason?**

~Maybe something like `REAUTHENTICATION_REQUIRED` ?~

No retry reason, since we've landed on not encouraging retries.

chvck replied:
Should we add this to a spec somewhere? 

> Maybe something like REAUTHENTICATION_REQUIRED ?

What's Java using now? (Sounds like Java already happens this status code?)

dnault replied:
Java currently throws a generic `CouchbaseException` with response status `UNKNOWN`. The error context mentions `AUTH_STALE`, but you have to look for it:

```
Exception in thread "main" com.couchbase.client.core.error.CouchbaseException: GetRequest failed with unexpected status code UNKNOWN {"completed":true,"coreId":"0x9fabcefe00000001","idempotent":true,"lastChannelId":"9FABCEFE00000001/000000001DE3879A","lastDispatchedFrom":"192.168.106.1:49512","lastDispatchedTo":"192.168.106.130:11207","requestId":12,"requestType":"GetRequest","retried":0,"service":{"bucket":"travel-sample","collection":"_default","documentId":"airline_10","errorCode":{"description":"Reauthentication required","name":"AUTH_STALE"},"opaque":"0x2f","scope":"_default","type":"kv","vbucket":105},"status":"UNKNOWN","timeoutMs":2500,"timings":{"dispatchMicros":1711,"totalDispatchMicros":1711,"totalServerMicros":0,"totalMicros":10065,"serverMicros":0}}
```

dnault replied:
I updated [this comment above](https://github.com/couchbaselabs/sdk-rfcs/pull/154/files#r2543341468) to reflect the latest plan: fail on `AUTH_STALE`. 

The failure message should give some indication that the request failed because the user credentials expired. An "AUTH_STALE" message in the error context is probably sufficient.

dnault committed:
Added a "Stale credentials" section that basically says,
"If you get an `AUTH_STALE` response, close the connection
and throw an exception."

Updated the "Credential rotation" section:

- SDK should prohibit user from switching authenticator types.

- `setAuthenticator()` either creates new connections, or reauthenticates
  existing connections (sometimes, depending on authenticator type;
  please re-read this section for all the gory details).
